### ISSUE TITLE
Incorrect DPR when launched in a second monitor

### ISSUE THREAD

**Issue by noritada** [alacritty/alacritty#4327]
Hi,

I encountered phenomena similar to #71.

When I launch the app from a second monitor with DPR 1, it displays the half-sized display in a correct size window. When I move the window to the main monitor with DPR 2, then it displays correctly, and when I move it again to the second monitor, it also displays correctly, although the log reports that it changed DPR 1 -> DPR 2 -> DPR 1.

On the other hand, when I launch alacritty from a primary builtin monitor with DPR 2, it works fine.

(It seems that "DPR 1" has already been detected before the display is drawn, but the display does not reflect it. I have checked the code and tried some changes, but cannot figure out how to run something like `terminal.resize(size_info)` inside the initialization process. What changes shall I try?)

| Action | Monitor   | Monitor DPR | Logged DPR | Expected DPR | Actual DPR |
|--------|-----------|---|---|---|---|
| Launch | secondary | 1 | 1 | 1 | 2? |
| Move   | primary   | 2 | 2 | 2 | 2  |
| Move   | secondary | 1 | 1 | 1 | 1  |

| Action | Monitor   | Monitor DPR | Logged DPR | Expected DPR | Actual DPR |
|--------|-----------|---|---|---|---|
| Launch | primary   | 2 | 2 | 2 | 2  |
| Move   | secondary | 1 | 1 | 1 | 1  |
| Move   | primary   | 2 | 2 | 2 | 2  |

![ss 2020-10-21 23 02 21（2）](https://user-images.githubusercontent.com/20850/96739281-922b6000-13fa-11eb-873a-72723fc9e35c.png)

### System

OS: macOS High Sierra (10.13.6)
Version: alacritty `0.6.0-dev (21eb84f)` (built from commit 21eb84fc5e0e649f53e3785957fa127ee9a78688)

### Logs

secondary (1) -> primary (2) -> secondary (1):
```
% /Applications/Alacritty.app/Contents/MacOS/alacritty -vv
[2020-10-21 23:02:18.127642000] [DEBUG] [alacritty] Using environment locale: ja_JP.UTF-8
Created log file at "/var/folders/wp/x1kk7wts6v3g_g1fy3zwp5c40000gn/T/Alacritty-13668.log"
[2020-10-21 23:02:18.128555000] [INFO ] [alacritty] Welcome to Alacritty
[2020-10-21 23:02:18.128596000] [INFO ] [alacritty] Configuration files loaded from:
[2020-10-21 23:02:18.128620000] [INFO ] [alacritty]   "/Users/noritada/.alacritty.yml"
[2020-10-21 23:02:18.199640000] [DEBUG] [alacritty] Estimated DPR: 2
[2020-10-21 23:02:18.199720000] [DEBUG] [alacritty] Estimated window size: None
[2020-10-21 23:02:18.199752000] [DEBUG] [alacritty] Estimated cell size: 13 x 29
[2020-10-21 23:02:18.251962000] [INFO ] [alacritty] Device pixel ratio: 1
[2020-10-21 23:02:18.256573000] [INFO ] [alacritty] Initializing glyph cache...
[2020-10-21 23:02:18.276219000] [INFO ] [alacritty] ... finished initializing glyph cache in 0.019514376s
[2020-10-21 23:02:18.276342000] [INFO ] [alacritty] Cell size: 6 x 15
[2020-10-21 23:02:18.276383000] [INFO ] [alacritty] Padding: 0 x 0
[2020-10-21 23:02:18.276835000] [INFO ] [alacritty] PTY dimensions: Line(40) x Column(133)
[2020-10-21 23:02:18.281977000] [INFO ] [alacritty] Initialisation complete
[2020-10-21 23:02:25.912539000] [INFO ] [alacritty] Font size changed to Size(22) with DPR of 2
[2020-10-21 23:02:25.917190000] [INFO ] [alacritty] Cell size: 13 x 29
[2020-10-21 23:02:25.917278000] [DEBUG] [alacritty_terminal] New num_cols is 123 and num_lines is 41
[2020-10-21 23:02:25.923895000] [INFO ] [alacritty] Padding: 0 x 0
[2020-10-21 23:02:25.923976000] [INFO ] [alacritty] Width: 1600, Height: 1200
[2020-10-21 23:02:38.556535000] [INFO ] [alacritty] Font size changed to Size(22) with DPR of 1
[2020-10-21 23:02:38.559250000] [INFO ] [alacritty] Cell size: 6 x 15
[2020-10-21 23:02:38.559313000] [DEBUG] [alacritty_terminal] New num_cols is 133 and num_lines is 40
[2020-10-21 23:02:38.565677000] [INFO ] [alacritty] Padding: 0 x 0
[2020-10-21 23:02:38.565782000] [INFO ] [alacritty] Width: 800, Height: 600
[2020-10-21 23:02:55.432480000] [INFO ] [alacritty] Goodbye
Deleted log file at "/var/folders/wp/x1kk7wts6v3g_g1fy3zwp5c40000gn/T/Alacritty-13668.log"
```

primary (2) -> secondary (1) -> primary (2):
```
% /Applications/Alacritty.app/Contents/MacOS/alacritty -vv
[2020-10-21 23:04:11.113045000] [DEBUG] [alacritty] Using environment locale: ja_JP.UTF-8
Created log file at "/var/folders/wp/x1kk7wts6v3g_g1fy3zwp5c40000gn/T/Alacritty-13726.log"
[2020-10-21 23:04:11.113213000] [INFO ] [alacritty] Welcome to Alacritty
[2020-10-21 23:04:11.113239000] [INFO ] [alacritty] Configuration files loaded from:
[2020-10-21 23:04:11.113259000] [INFO ] [alacritty]   "/Users/noritada/.alacritty.yml"
[2020-10-21 23:04:11.156174000] [DEBUG] [alacritty] Estimated DPR: 2
[2020-10-21 23:04:11.156247000] [DEBUG] [alacritty] Estimated window size: None
[2020-10-21 23:04:11.156274000] [DEBUG] [alacritty] Estimated cell size: 13 x 29
[2020-10-21 23:04:11.199306000] [INFO ] [alacritty] Device pixel ratio: 2
[2020-10-21 23:04:11.203885000] [INFO ] [alacritty] Initializing glyph cache...
[2020-10-21 23:04:11.225220000] [INFO ] [alacritty] ... finished initializing glyph cache in 0.0209997s
[2020-10-21 23:04:11.225399000] [INFO ] [alacritty] Cell size: 13 x 29
[2020-10-21 23:04:11.225447000] [INFO ] [alacritty] Padding: 0 x 0
[2020-10-21 23:04:11.225823000] [INFO ] [alacritty] PTY dimensions: Line(41) x Column(123)
[2020-10-21 23:04:11.229156000] [INFO ] [alacritty] Initialisation complete
[2020-10-21 23:04:17.884078000] [INFO ] [alacritty] Font size changed to Size(22) with DPR of 1
[2020-10-21 23:04:17.890079000] [INFO ] [alacritty] Cell size: 6 x 15
[2020-10-21 23:04:17.890158000] [DEBUG] [alacritty_terminal] New num_cols is 133 and num_lines is 40
[2020-10-21 23:04:17.895593000] [INFO ] [alacritty] Padding: 0 x 0
[2020-10-21 23:04:17.895691000] [INFO ] [alacritty] Width: 800, Height: 600
[2020-10-21 23:04:26.658052000] [INFO ] [alacritty] Font size changed to Size(22) with DPR of 2
[2020-10-21 23:04:26.661061000] [INFO ] [alacritty] Cell size: 13 x 29
[2020-10-21 23:04:26.661145000] [DEBUG] [alacritty_terminal] New num_cols is 123 and num_lines is 41
[2020-10-21 23:04:26.673038000] [INFO ] [alacritty] Padding: 0 x 0
[2020-10-21 23:04:26.673162000] [INFO ] [alacritty] Width: 1600, Height: 1200
[2020-10-21 23:04:38.168251000] [INFO ] [alacritty] Goodbye
Deleted log file at "/var/folders/wp/x1kk7wts6v3g_g1fy3zwp5c40000gn/T/Alacritty-13726.log"
```

**Comment by chrisduerr** [comment]
Is this a regression or has this always been a problem for you?

**Comment by noritada** [comment]
I am sorry but I have started this good terminal emulator recently, I have not used previous versions. Shall I build and test previous versions? If so, are there any versions that should be checked?

**Comment by chrisduerr** [comment]
The latest stable version (0.5.0) would be good.

**Comment by noritada** [comment]
OK, I checked with `0.5.0 (a6681e3)` (both downloaded one and built one) and got the same result.
So, this is not a regression after 0.5.0.

**Comment by chrisduerr** [comment]
Does https://github.com/alacritty/alacritty/issues/3695 sound familiar?

**Comment by chrisduerr** [comment]
Sounds like the problem is with the estimated DPR not matching the actual DPR.

**Comment by noritada** [comment]
> Does #3695 sound familiar?

It does not seem to be.

> Sounds like the problem is with the estimated DPR not matching the actual DPR.

I think so since the problem does not occur when they match.

> The screenshot is a tad low when it comes to resolution, I'd assume that the problem is that no matter how often you press enter, you don't get below a certain point in the terminal?

You are almost right, except that I cannot reach the bottom. I cannot reach the top of the window.

![ss 2020-10-22 1 38 04](https://user-images.githubusercontent.com/20850/96751882-b4c47580-1408-11eb-81b0-818cfc3ef75b.png)

**Comment by noritada** [comment]
I have found that when I resize the window in the same monitor, then the display expand to the window size (i.e. correct size). So, window events can update DPR correctly. The problem can be observed only after launch and before events issued.

**Comment by chrisduerr** [comment]
Could you try this patch:

```diff
diff --git a/alacritty/src/display.rs b/alacritty/src/display.rs
index 0c5948f..8a6cd22 100644
--- a/alacritty/src/display.rs
+++ b/alacritty/src/display.rs
@@ -206,9 +206,6 @@ impl Display {
 
         info!("Device pixel ratio: {}", window.dpr);
 
-        // get window properties for initializing the other subsystems.
-        let viewport_size = window.inner_size();
-
         // Create renderer.
         let mut renderer = QuadRenderer::new()?;
 
@@ -226,6 +223,7 @@ impl Display {
         }
 
         let padding = config.ui_config.window.padding(window.dpr);
+        let viewport_size = window.inner_size();
 
         // Create new size with at least one column and row.
         let size_info = SizeInfo::new(
@@ -240,6 +238,7 @@ impl Display {
 
         info!("Cell size: {} x {}", cell_width, cell_height);
         info!("Padding: {} x {}", size_info.padding_x(), size_info.padding_y());
+        info!("Width: {}, Height: {}", size_info.width(), size_info.height());
 
         // Update OpenGL projection.
         renderer.resize(&size_info);
```

It shouldn't fix anything, but it will make your `alacritty -vv` log more useful. I'm curious only about the broken case of 1 -> 2 -> 1.

**Comment by noritada** [comment]
Log of the broken case (1 -> 2 -> 1):
```
% ./target/release/osx/Alacritty.app/Contents/MacOS/alacritty -vv
[2020-10-22 02:03:50.499979000] [DEBUG] [alacritty] Using environment locale: ja_JP.UTF-8
Created log file at "/var/folders/wp/x1kk7wts6v3g_g1fy3zwp5c40000gn/T/Alacritty-17338.log"
[2020-10-22 02:03:50.500444000] [INFO ] [alacritty] Welcome to Alacritty
[2020-10-22 02:03:50.500489000] [INFO ] [alacritty] Configuration files loaded from:
[2020-10-22 02:03:50.500517000] [INFO ] [alacritty]   "/Users/noritada/.alacritty.yml"
[2020-10-22 02:03:50.554264000] [DEBUG] [alacritty] Estimated DPR: 2
[2020-10-22 02:03:50.554323000] [DEBUG] [alacritty] Estimated window size: None
[2020-10-22 02:03:50.554346000] [DEBUG] [alacritty] Estimated cell size: 13 x 29
[2020-10-22 02:03:50.597617000] [INFO ] [alacritty] Device pixel ratio: 1
[2020-10-22 02:03:50.601297000] [INFO ] [alacritty] Initializing glyph cache...
[2020-10-22 02:03:50.617536000] [INFO ] [alacritty] ... finished initializing glyph cache in 0.016116884s
[2020-10-22 02:03:50.617662000] [INFO ] [alacritty] Cell size: 6 x 15
[2020-10-22 02:03:50.617707000] [INFO ] [alacritty] Padding: 0 x 0
[2020-10-22 02:03:50.617744000] [INFO ] [alacritty] Width: 800, Height: 600
[2020-10-22 02:03:50.618096000] [INFO ] [alacritty] PTY dimensions: Line(40) x Column(133)
[2020-10-22 02:03:50.624074000] [INFO ] [alacritty] Initialisation complete
[2020-10-22 02:03:58.545334000] [INFO ] [alacritty] Font size changed to Size(22) with DPR of 2
[2020-10-22 02:03:58.550258000] [INFO ] [alacritty] Cell size: 13 x 29
[2020-10-22 02:03:58.550356000] [DEBUG] [alacritty_terminal] New num_cols is 123 and num_lines is 41
[2020-10-22 02:03:58.563989000] [INFO ] [alacritty] Padding: 0 x 0
[2020-10-22 02:03:58.564099000] [INFO ] [alacritty] Width: 1600, Height: 1200
[2020-10-22 02:04:04.512175000] [INFO ] [alacritty] Font size changed to Size(22) with DPR of 1
[2020-10-22 02:04:04.515809000] [INFO ] [alacritty] Cell size: 6 x 15
[2020-10-22 02:04:04.515917000] [DEBUG] [alacritty_terminal] New num_cols is 133 and num_lines is 40
[2020-10-22 02:04:04.520343000] [INFO ] [alacritty] Padding: 0 x 0
[2020-10-22 02:04:04.520468000] [INFO ] [alacritty] Width: 800, Height: 600
[2020-10-22 02:04:10.470507000] [INFO ] [alacritty] Goodbye
Deleted log file at "/var/folders/wp/x1kk7wts6v3g_g1fy3zwp5c40000gn/T/Alacritty-17338.log"
```

It seems to be the same as the initial one.

**Comment by chrisduerr** [comment]
> It seems to be the same as the initial one except that "Width: 800, Height: 600" added.

Yeah, that's the goal.

I really don't see what would be wrong here. It's detecting everything correctly. The window size is right, the cell size is right, the DPR is right. Do you have anything noteworthy about your setup that might be a bit unusual?

**Comment by noritada** [comment]
I don't have anything as of now. I'd like to check whether it can be reproduced in environments of macOS users around me.

**Comment by chrisduerr** [comment]
> OS: macOS High Sierra (10.13.6)

This strikes me as odd. That's severely outdated, isn't it?

**Comment by noritada** [comment]
Yes, that is why I want to check other environments. Although security update was provided for 10.13 1 month ago (on Sep. 24), the first release of 10.13 was 3 years ago (on Sep 25, 2017).

**Comment by noritada** [comment]
When I was reviewing past issues, I found that this issue seems to be a duplicate of #1631. However, since that issue was also reported for macOS High Sierra, I got no updates.

I will continue to find other environments around me where this issue occurs.

**Comment by chrisduerr** [comment]
I knew this issue sounded familiar, thanks. Though as previously stated I think this might just be fixed by updating macOS.

**Comment by noritada** [comment]
I have upgraded my macOS from High Sierra (10.13.6) to Mojave (10.14.6) and got a correct display with the same binary as what I used to see incorrect display on 10.13.6. So, confirmed that this is only an issue on 10.13, which is now 3 years old. Sorry for the disturbance. Thanks!
